bin/ch/Debugger.cpp

```diff
@@ -0,0 +1,526 @@
+
+        IfJsErrorFailLogAndRet(ChakraRTInterface::JsDiagEvaluate(str, stackFrameIndex, &result));
+    }
+
+    return result;
+}
+
+Debugger::Debugger(JsRuntimeHandle runtime)
+{
+    this->m_runtime = runtime;
+    this->m_context = JS_INVALID_REFERENCE;
+    this->isDetached = true;
+}
+
+Debugger::~Debugger()
+{
+    this->m_runtime = JS_INVALID_RUNTIME_HANDLE;
+}
+
+Debugger * Debugger::GetDebugger(JsRuntimeHandle runtime)
+{
+    if (Debugger::debugger == nullptr)
+    {
+        Debugger::debugger = new Debugger(runtime);
+        Debugger::debugger->Initialize();
+    }
+
+    return Debugger::debugger;
+}
+
+void Debugger::CloseDebugger()
+{
+    if (Debugger::debugger != nullptr)
+    {
+        delete Debugger::debugger;
+        Debugger::debugger = nullptr;
+    }
+}
+
+JsRuntimeHandle Debugger::GetRuntime()
+{
+    return Debugger::debugger != nullptr ? Debugger::debugger->m_runtime : JS_INVALID_RUNTIME_HANDLE;
+}
+
+bool Debugger::Initialize()
+{
+    // Create a new context and run dbgcontroller.js in that context
+    // setup dbgcontroller.js callbacks
+
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsCreateContext(this->m_runtime, &this->m_context));
+
+    AutoRestoreContext autoRestoreContext(this->m_context);
+
+    JsValueRef globalFunc = JS_INVALID_REFERENCE;
+
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsParseScriptWithAttributes(controllerScript, JS_SOURCE_CONTEXT_NONE, _u("DbgController.js"), JsParseScriptAttributeLibraryCode, &globalFunc));
+
+    JsValueRef undefinedValue;
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsGetUndefinedValue(&undefinedValue));
+
+    JsValueRef args[] = { undefinedValue };
+    JsValueRef result = JS_INVALID_REFERENCE;
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsCallFunction(globalFunc, args, _countof(args), &result));
+
+    JsValueRef globalObj = JS_INVALID_REFERENCE;
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsGetGlobalObject(&globalObj));
+
+    JsPropertyIdRef hostDebugObjectPropId;
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsGetPropertyIdFromName(_u("hostDebugObject"), &hostDebugObjectPropId));
+
+    JsPropertyIdRef hostDebugObject;
+    IfJsrtErrorFailLogAndRetFalse(ChakraRTInterface::JsGetProperty(globalObj, hostDebugObjectPropId, &hostDebugObject));
+
+    this->InstallDebugCallbacks(hostDebugObject);
+
+    if (!WScriptJsrt::Initialize())
+    {
+        return false;
+    }
+
+    if (HostConfigFlags::flags.dbgbaselineIsEnabled)
+    {
+        this->SetBaseline();
+    }
+
+    if (HostConfigFlags::flags.InspectMaxStringLengthIsEnabled)
+    {
+        this->SetInspectMaxStringLength();
+    }
+
+    return true;
+}
+
+bool Debugger::InstallDebugCallbacks(JsValueRef hostDebugObject)
+{
+    HRESULT hr = S_OK;
+    IfFalseGo(WScriptJsrt::InstallObjectsOnObject(hostDebugObject, _u("JsDiagGetSource"), Debugger::JsDiagGetSource));
```

Comment:
I wonder if you need to specify JsDiag here as you are already under Debugger container.